// irulan-single-page-application/index.js
var fs = require("fs");
var path = require("path");
module.app = {};

module.exports = SinglePageApplication;
module.exports.IndexPage = IndexPage;
module.exports.ErrorPage = ErrorPage;
module.exports.InjectClientScripts = InjectClientScripts;

// Constructor
function SinglePageApplication(_app, router) {
  //console.log(module.app);
  //console.log(this.app);
  app = module.app = _app;
  //console.log(this.app);

  // Inject the client side scripts we'll need
  app.ClientScripts['single page application'].push({
    name: 'angular.min.js',
    online_url: '//ajax.googleapis.com/ajax/libs/angularjs/1.4.8/angular.min.js',
    offline_path: app.Settings['cachedJsDir'] + '/angular.min.js'
  })
  app.ClientScripts['single page application'].push({
    name: 'angular-sanitize.min.js',
    online_url: '//ajax.googleapis.com/ajax/libs/angularjs/1.4.8/angular-sanitize.min.js',
    offline_path: app.Settings['cachedJsDir'] + '/angular-sanitize.min.js'
  })
  app.ClientScripts['single page application'].push({
    name: 'angular-route.min.js',
    online_url: '//cdnjs.cloudflare.com/ajax/libs/angular.js/1.4.8/angular-route.min.js',
    offline_path: app.Settings['cachedJsDir'] + '/angular-route.min.js'
  })
  app.ClientScripts['single page application'].push({
    name: 'jquery.min.js',
    online_url: '//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js',
    offline_path: app.Settings['cachedJsDir'] + '/jquery.min.js'
  })
  app.ClientScripts['single page application'].push({
    name: 'bootstrap.min.js',
    online_url: '//maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js',
    offline_path: app.Settings['cachedJsDir'] + '/bootstrap.min.js'
  })
  app.ClientScripts['single page application'].push({
    name: 'angular-resource.min.js',
    online_url: '//cdnjs.cloudflare.com/ajax/libs/angular.js/1.4.8/angular-resource.min.js',
    offline_path: app.Settings['cachedJsDir'] + '/angular-resource.min.js'
  })
  app.ClientScripts['single page application'].push({
    name: 'app.js',
    generate: function() {
      console.log('app.js needs certain dynamic adjustments depending on which plugins are enabled.');
      return ""
    }
  })
  app.ClientScripts['single page application'].push({
    name: 'components.navbar.js',
    offline_path: app.Settings['clientJsDir'] + '/components.navbar.js'
  })
  app.ClientScripts['single page application'].push({
    name: 'components.pages.js',
    offline_path: app.Settings['clientJsDir'] + '/components.pages.js'
  })

  // Register the fact that we will be handling requests for "/" (i.e.
  //  index.html)
  router.get("/", function (req, res) {
    IndexPage(req, res);
  });

  // We also want a chance to provide any ".js" files the client requests
  router.get(/\.js$/, function (req, res, next) {
    console.log(".js file request: " + req._parsedUrl.path)
    SendScript(req, res, next);
  });
}

// Take all of the objects in "Settings['client scripts']" and generate
//  the HTML that needs to be injected into the HTML page.
function InjectClientScripts() {
  var app = module.app;

  var html = '';
  for (index in app.ClientScripts['single page application']) {
    script = app.ClientScripts['single page application'][index];
    if (("generate" in script)) {
      html += '<script src="/' + script.name + '"></script>\n';
    }
    else if (app.Settings['offline mode']) {
      if (!("offline_path" in script)) {
        if (("online_url" in script)) {
          console.log('We are in offline mode but no offline path for "' + script.name + '" is provided.');
          html += '<script src="' + script.online_url + '"></script>\n';
        }
        else {
          console.log('Neither an online_url or offline_path was available for "' + script.name + '".');
        }
      }
      else {
        html += '<script src="/' + script.name + '"></script>\n';
      }
    }
    else if (("online_url" in script)) {
      html += '<script src="' + script.online_url + '"></script>\n';
    }
    else if (("offline_path" in script)) {
      html += '<script src="/' + script.name + '"></script>\n';
    }
    else {
      console.log('Neither an online_url or offline_path was available for "' + script.name + '".');
    }
  }
  return html;
}

function SendScript(req, res, next) {
  var app = module.app;

  //TODO: This always goes to "next()".  Finish writing it!
  requestedScript = path.basename(req._parsedUrl.path);
  if (requestedScript == 'all.js') {
    //TODO: merge all of the script files into one and send
  }
  else {
    for (index in app.ClientScripts['single page application']) {
      script = app.ClientScripts['single page application'][index];
      if (script.name == requestedScript) {
        if (("generate" in script)) {
          console.log('Should generate file from using generate() function');
        }
        else {
          console.log('Should send file from ' + script.offline_path);
        }
      }
    }
  }
  next();
}

function IndexPage(req, res) {
  var app = module.app;

  console.log("IndexPage");
  res.sendFile(app.Settings['views directory'] + "index.html");
}

function ErrorPage(req, res, error_num, error_text) {
  var app = module.app;

  res.sendFile(app.Settings['views directory'] + "404.html");
}
